tianshou/policy/imitation/bcq.py

```diff
@@ -0,0 +1,326 @@
+        self.lmbda = lmbda
+        self.device = device
+
+    def train(self, mode: bool = True) -> "BCQPolicy":
+        """Set the module in training mode, except for the target network."""
+        self.training = mode
+        self.actor.train(mode)
+        self.critic1.train(mode)
+        self.critic2.train(mode)
+        return self
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data."""
+        # state: None, input: "obs"
+        # There is "obs" in the Batch
+        # obs: 10 groups. Each group has a state. shape: (10, state_dim)
+        obs_group = torch.FloatTensor(batch["obs"]).to(self.device)
+
+        act = []
+        with torch.no_grad():
+            for obs in obs_group:
+                # now obs is (state_dim)
+                obs = (obs.reshape(1, -1)).repeat(100, 1)
+                # now obs is (100, state_dim)
+
+                # decode(obs) generates action and actor perturbs it
+                action = self.actor(obs, self.vae.decode(obs))
+                # now action is (100, action_dim)
+                q1 = self.critic1(obs, action)
+                # q1 is (100, 1)
+                ind = q1.argmax(0)
+                act.append(action[ind].cpu().data.numpy().flatten())
+        act = np.array(act)
+        return Batch(act=act)
+
+    def sync_weight(self) -> None:
+        """Soft-update the weight for the target network."""
+        for param, target_param in \
+                zip(self.critic1.parameters(), self.critic1_target.parameters()):
+            target_param.data.copy_(
+                self.tau * param.data + (1 - self.tau) * target_param.data
+            )
+        for param, target_param in \
+                zip(self.critic2.parameters(), self.critic2_target.parameters()):
+            target_param.data.copy_(
+                self.tau * param.data + (1 - self.tau) * target_param.data
+            )
+
+        for param, target_param in \
+                zip(self.actor.parameters(), self.actor_target.parameters()):
+            target_param.data.copy_(
+                self.tau * param.data + (1 - self.tau) * target_param.data
+            )
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        # batch: obs, act, rew, done, obs_next. (numpy array)
+        # (batch_size, state_dim)
+        obs = torch.FloatTensor(batch["obs"]).to(device=self.device)
+        # (batch_size, action_dim)
+        act = torch.FloatTensor(batch["act"]).to(device=self.device)
+        # (batch_size)
+        rew = torch.FloatTensor(batch["rew"]).to(device=self.device)
+        # (batch_size)
+        done = torch.IntTensor(batch["done"]).to(device=self.device)
+        # (batch_size, state_dim)
+        obs_next = torch.FloatTensor(batch["obs_next"]).to(device=self.device)
```

Comment:
same, for conversion you can refer to other policy's approaches instead of manually handling this (e.g., what if obs is not a numpy array?)